clustermesh-apiserver/users_mgmt.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+const (
+	usersMgmtCtrl = "clustermesh-users-management"
+)
+
+var usersManagementCell = cell.Module(
+	"clustermesh-users-management",
+	"ClusterMesh Etcd Users Management",
+
+	kvstore.GlobalUserMgmtClientPromiseCell,
+	cell.Config(UsersManagementConfig{}),
+	cell.Invoke(registerUsersManager),
+)
+
+type UsersManagementConfig struct {
+	ClusterUsersEnabled    bool
+	ClusterUsersConfigPath string
+}
+
+func (UsersManagementConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("cluster-users-enabled", false,
+		"Enable the management of etcd users for remote clusters")
+	flags.String("cluster-users-config-path", "/var/lib/cilium/etcd-config/users.yaml",
+		"The path of the config file with the list of remote cluster users")
+}
+
+type usersConfigFile struct {
+	Users []struct {
+		Name string `yaml:"name"`
+		Role string `yaml:"role"`
+	} `yaml:"users"`
+}
+
+type usersManager struct {
+	UsersManagementConfig
+
+	client        kvstore.BackendOperationsUserMgmt
+	clientPromise promise.Promise[kvstore.BackendOperationsUserMgmt]
+
+	manager *controller.Manager
+	users   map[string]string
+
+	stop chan struct{}
+	wg   sync.WaitGroup
+}
+
+func registerUsersManager(lc hive.Lifecycle, cfg UsersManagementConfig, clientPromise promise.Promise[kvstore.BackendOperationsUserMgmt]) error {
+	if !cfg.ClusterUsersEnabled {
+		log.Info("etcd users management disabled")
+		return nil
+	}
+
+	manager := usersManager{
+		UsersManagementConfig: cfg,
+		clientPromise:         clientPromise,
+
+		manager: controller.NewManager(),
+		users:   make(map[string]string),
+
+		stop: make(chan struct{}),
+	}
+
+	lc.Append(&manager)
+	return nil
+}
+
+func (us *usersManager) Start(hive.HookContext) error {
+	log.WithField(logfields.Path, us.ClusterUsersConfigPath).
+		Info("Starting managing etcd users based on configuration")
+
+	configWatcher, err := fswatcher.New([]string{us.ClusterUsersConfigPath})
+	if err != nil {
+		log.WithError(err).Error("Unable to setup config watcher")
+		return fmt.Errorf("unable to setup config watcher: %w", err)
+	}
+
+	us.manager.UpdateController(usersMgmtCtrl, controller.ControllerParams{
+		Context: context.Background(),
+		DoFunc:  us.sync,
+	})
+
+	us.wg.Add(1)
+	go func() {
+		defer us.wg.Done()
+
+		for {
+			select {
+			case <-configWatcher.Events:
+				us.manager.TriggerController(usersMgmtCtrl)
+			case err := <-configWatcher.Errors:
+				log.WithError(err).WithField(logfields.Path, us.ClusterUsersConfigPath).
+					Warning("Error encountered while watching file with fsnotify")
+			case <-us.stop:
+				log.Info("Closing")
+				configWatcher.Close()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (us *usersManager) Stop(hive.HookContext) error {
+	log.WithField(logfields.Path, us.ClusterUsersConfigPath).
+		Info("Stopping managing etcd users based on configuration")
+
+	us.manager.RemoveAllAndWait()
```

Comment:
Why is this necessary?

Reply:
It ensures that the controller which takes care of synchronizing the users gets stopped when Stop is called. And it waits for it to terminate as per the second guideline here: https://docs.cilium.io/en/latest/contributing/development/hive/#guidelines (if I got your question correctly).